tf2/src/buffer_core.cpp

```diff
@@ -1624,39 +1544,37 @@ void BufferCore::_chainAsVector(
   tf2::TF2Error retval = walkToTopParent(
     accum, source_time, fixed_id, source_id, &error_string,
     &source_frame_chain);
+  switch (retval) {
+    case tf2::TF2Error::NO_ERROR:
+      break;
+    case tf2::TF2Error::CONNECTIVITY_ERROR:
+      throw ConnectivityException(error_string);
+    case tf2::TF2Error::EXTRAPOLATION_ERROR:
+      throw ExtrapolationException(error_string);
+    case tf2::TF2Error::LOOKUP_ERROR:
+      throw LookupException(error_string);
+    default:
+      throw std::runtime_error("Unknown error code: " + std::to_string(static_cast<int>(retval)));
+  }
 
-  if (retval != tf2::TF2Error::NO_ERROR) {
+  if (source_time != target_time) {
+    std::vector<CompactFrameID> target_frame_chain;
+    retval = walkToTopParent(
+      accum, target_time, target_id, fixed_id, &error_string,
+      &target_frame_chain);
     switch (retval) {
+      case tf2::TF2Error::NO_ERROR:
```

Comment:
I believe that this is a performance optimization to have the if and the switch separate. It would be good to see that this doesn't effect the performance benchmark tests before merging this. Reading about time complexity of switch it seems like this should be able to be optimized by the compiler but these flows were highly optimized by Josh at one point based on benchmark tests.

Reply:
So I built an example program and looked at the assembly.

When building without any optimizations, it looks like you are correct; wrapping the `switch` statement in an `if` saves one `jmp`.  Interestingly, when turning on `-O3`, it is the opposite; using the `switch` statement only takes 3 instructions, while wrapping it with an `if` takes 5.  These results are, of course, highly architecture and compiler specific.

Taking a step back, though, this difference is unlikely to show up in most benchmarks.  You are probably going to be spending your time elsewhere, so the difference here of a few instructions either way doesn't seem that relevant.  I'd rather go for code clarity here.

Reply:
I ended up going back to the original code, just to keep the diff here down.  I still think that the difference in speed isn't worth it, but we can come back to that another time.